distributed/worker.py

```diff
@@ -3089,7 +3091,10 @@ async def gather_dep(
                         self.batched_stream.send(
                             {"op": "missing-data", "errant_worker": worker, "key": d}
                         )
-                        recommendations[ts] = "fetch" if ts.who_has else "missing"
+                        if ts.who_has:
+                            recommendations[ts] = "fetch"
+                        elif ts.state not in ("released", "memory"):
+                            recommendations[ts] = "missing"
```

Comment:
Could the task be in memory or missing even if `who_has` is non empty?

Guess I'm wondering why this isn't:
```suggestion
                        if ts.state not in ("released", "memory"):
                            recommendations[ts] = "fetch" if ts.who_has else "missing"
```

For that matter, could `ts` be errored? If `ts` was stolen to this worker (and executed) in the time since the `gather_dep` started, and execution failed? What about other terminal states?

I feel like an passlist here (likely fetch/flight) still would make more sense. If some action has occurred in the meantime since `gather_dep` started which has changed the task's state to something non-dependency-fetch-y, that probably takes precedence over re-fetching it right? I'd be curious if/why an passlist doesn't work.

Reply:
I'm happy to make a change like what is above because it seems safe relative to what is here already.

When I implemented the allowlist / passlist tests failed, so our reasoning here is not yet correct.   I think that we should absolutely pursue this, but that we should handle that separately.  I'd like to get this in for a release and I don't want to increase scope much beyond "avoid released/memory -> missing transitions".  I totally get that we might want to do more eventually, but I don't think that we need to do more in order to do this.

Reply:
> I'm happy to make a change like what is above because it seems safe relative to what is here already.

Wait, no, it looks like that would exclude more paths.  I'm really looking for a minimal set of changes that avoid the deadlock for right now.